### PR TITLE
Rework meson-vcs-tag.sh

### PR DESCRIPTION
**Pull request by daandemeyer**
git describe was not made for a git rebase workflow. git describe's output looks as follows: "v255-1234-ga1d5f3" where the first component is the nearest tag, the second component is the number of commits since that tag and the last component is the short commit sha of the current commit.

If git rebase is not used, the git describe output for later commits will always sort higher than the git describe output of earlier commits when the output is used as a version. However, if git rebase is used, the commit sha of an existing commit might change and suddenly sort lower than the previous sha of the same commit, which means the corresponding version will also sort lower.

To fix this problem, we stop using git describe in meson-vcs-tag.sh. Instead, we opt to use a "<version>.<date>.<sha>" format which guarantees that later builds will always sort higher than earlier builds, while keeping the useful information of the current commit hash in the version. We also keep appending the version with a caret if the current working tree is dirty.

The new format looks as follows:

➜  systemd git:(versioning) tools/meson-vcs-tag.sh 256~devel.20240216132005.gf508f3b785^

### CODE REVIEW

**Comment by keszybz** [comment]
Let's try this. I think it should work. Maybe we'll discover that some tweaks are needed, but it's easier to discover those after merging.

ci/semaphoreci/pr: Debian autopkgtest (LXC) failed during installation, trying to remove some files. Seems to be an install fuckup, but not relate to the version string.

**Review by keszybz — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
@keszybz Please take another look, the previous approach caused way too many targets to always be rebuilt as the version was changing every single build even if no input files changed.

My new idea is to make the package release a timestamp and to not propagate the package release into meson from the packaging specs if we're doing an upstream build. So this PR now splits up version-tag into version-tag and vcs-tag so that an extended version can still be passed while still having meson calculate the current git commit.

In the fedora spec we'd do something like the following:

```
%if %{without upstream}
Version:        255.3
Release:        %autorelease
%else
# determine the build information from local checkout
Version:        %(tools/meson-vcs-tag.sh)
Release:        %(date '+%Y%m%d%H%M%S')%{?dist}
%endif

...

-Dversion-tag=%{version}-%[%{without upstream}?"%{release}":"1"]
```

File: meson.build [hunk 1]
```diff
@@ -1853,13 +1853,27 @@ xml_helper_py = find_program('tools/xml_helper.py')
...
+
+conf.set_quoted('VERSION_TAG', version_tag)
+
+vcs_tag = get_option('vcs-tag')
+if vcs_tag
+        version_h = vcs_tag(
+                input : 'src/version/version.h.in',
+                output : 'version.h',
+                fallback : '',
+                command : ['sh', '-c', 'echo " r$(git rev-parse --short HEAD)"'],
```
Comment (keszybz) [inline meson.build]: Hmm, I don't understand why that would work, since the git directory is not specified.

Comment (daandemeyer) [reply meson.build]: I think it automatically runs from the source directory.

Comment (keszybz) [reply meson.build]: Why "r"? We talked about "g" before.
`--abbrev=7` got lost. The default is 10 chars, and that seems a bit too much.
`--dirty=^` got lost. This is important to tell the developer that they're not actually testing the code that is described by that hash.

File: meson_options.txt [hunk 2]
```diff
@@ -2,9 +2,11 @@
 # SPDX-License-Identifier: LGPL-2.1-or-later
 
 option('version-tag', type : 'string',
-       description : 'override the git version string')
+       description : 'set the extended version string (defaults to project version)')
```
Comment (keszybz) [inline meson_options.txt:5]: "extended"? Isn't this actually the partial version string?

File: src/version/version.h.in [hunk 3]
```diff
@@ -7,4 +7,4 @@
  * - where a simplified machine-parsable form is more useful, for example
  *   pkgconfig files and version information written to binary files.
  */
-#define GIT_VERSION "@VCS_TAG@"
+#define GIT_VERSION VERSION_TAG "@VCS_TAG@"
```
Comment (keszybz) [inline src/version/version.h.in:10]: This runs together: `256~develrdeadbeef`.

**Review by keszybz — CHANGES_REQUESTED** [review]

**Comment by keszybz** [comment]
When executed **not** in a git repo, this says:
```
[40/2478] Generating version.h with a custom command
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```
and `#define GIT_VERSION VERSION_TAG "r"`.

**Comment by keszybz** [comment]
Thinking about this some more, just calling `git` and hoping for the best is not going to work if we are inside of another repository. For example, consider the case of `fedpkg local` done from inside of a dist-git repo. We do not want to use the commit info from dist-git, but that's what we're going to get, because the sources are checked out from a tarball without git metadata.

I *think* we may get away with just `-C .`, since it seems that meson calls the generator in the source directory always, as you said.

File: meson_options.txt (hunk 2 above)
Comment (daandemeyer) [reply meson_options.txt:5]: I figured the project version was the partial version string. The extended version is this one and the git version is the one with the commit information added.

**Comment by keszybz** [comment]
Computer says no:
```
[1/2377] Generating version.h with a custom command
fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

**Comment by DaanDeMeyer** [comment]
Added a check for a `.git` directory

**Comment by keszybz** [comment]
Seems to work fine.

**Pull request merged**